### PR TITLE
Update label_id in label-scan

### PR DESCRIPTION
**Pull request by raz-mon**
Update label_id of label in cached label-scan operations.

### CODE REVIEW

File: tests/flow/test_optimizations_plan.py [hunk 1]
```diff
@@ -426,3 +426,31 @@ def test27_optimize_cartesian_product_complex_filter_trees(self):
                     [0, 3],
                     [0, 3]]
         self.env.assertEqual(resultset, expected)
+
+    # Label_id of a label in a cached execution plan should be updated properly.
+    # The label with less nodes should be traversed first.
+    def test28_optimize_label_scan_cached_label_id(self):
```
Comment (swilly22) [inline tests/flow/test_optimizations_plan.py:432]: Test intro comment:
`in cases where a referred label doesn't exists, the UNKNOW_LABEL_ID is being cached
once the label is created we want to make sure that the UNKNOW_LABEL_ID is replaced with the actual label ID
this test illustrates this scenario by traversing from a non-existing label (populating our execution-plan cache) which afterwards is being created, once created we want to make sure the correct label ID is used.`

File: tests/flow/test_optimizations_plan.py [hunk 2]
```diff
@@ -426,3 +426,31 @@ def test27_optimize_cartesian_product_complex_filter_trees(self):
...
+        self.env.flush()
+
+        # Create node with label Q
+        graph.query("CREATE (n:Q {v: 1})")
+
+        # Make sure N is traversed first, as it has less nodes.
```
Comment (swilly22) [inline tests/flow/test_optimizations_plan.py:438]: Make sure N is traversed first, as it has no nodes. (none existing)

File: tests/flow/test_optimizations_plan.py [hunk 3]
```diff
@@ -426,3 +426,31 @@ def test27_optimize_cartesian_product_complex_filter_trees(self):
...
+        plan = graph.execution_plan("MATCH (n:N:Q) RETURN n.v")
+        self.env.assertIn("Node By Label Scan | (n:N)", plan)
+
+        # Add label `N` to only node in the graph
+        query = """MATCH (n:Q) SET n:N"""
+        graph.query(query)
+
+        # Make sure #nodes labeled as Q > #nodes labeled as N
+        graph.query("CREATE (n:Q {v: 1})")
```
Comment (swilly22) [inline tests/flow/test_optimizations_plan.py:447]: I don't think we need attributes `{v: 1}`

File: tests/flow/test_optimizations_plan.py [hunk 4]
```diff
@@ -426,3 +426,31 @@ def test27_optimize_cartesian_product_complex_filter_trees(self):
...
+
+        # Make sure N is traversed first, N is now associated with an ID
+        # |N| < |Q|
+        query = """MATCH (n:N:Q) RETURN n.v"""
```
Comment (swilly22) [inline tests/flow/test_optimizations_plan.py:451]: `MATCH (n:N:Q) RETURN count(n)`

File: tests/flow/test_optimizations_plan.py [hunk 5]
```diff
@@ -426,3 +426,31 @@ def test27_optimize_cartesian_product_complex_filter_trees(self):
...
                     [0, 3]]
         self.env.assertEqual(resultset, expected)
+
+    # Label_id of a label in a cached execution plan should be updated properly.
+    # The label with less nodes should be traversed first.
+    def test28_optimize_label_scan_cached_label_id(self):
+        self.env.flush()
+
+        # Create node with label Q
+        graph.query("CREATE (n:Q {v: 1})")
```
Comment (swilly22) [inline tests/flow/test_optimizations_plan.py:436]: no attribute

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/execution_plan/ops/op_node_by_label_scan.c [hunk 6]
```diff
@@ -24,13 +24,21 @@ static inline void NodeByLabelScanToString(const OpBase *ctx, sds *buf) {
...
 }
 
+static void _update_label_id(NodeByLabelScan *op) {
+	if(op->n.label_id != GRAPH_UNKNOWN_LABEL) return;
+
+	GraphContext *gc = QueryCtx_GetGraphCtx();
+	Schema *s = GraphContext_GetSchema(gc, op->n.label, SCHEMA_NODE);
+	if(s != NULL) op->n.label_id = Schema_GetID(s);
+}
+
```
Comment (swilly22) [inline src/execution_plan/ops/op_node_by_label_scan.c:34]: Please include a short comment describing the purpose and the reason for this function

File: src/execution_plan/ops/op_node_by_label_scan.c [hunk 7]
```diff
@@ -90,19 +98,17 @@ static OpResult NodeByLabelScanInit(OpBase *opBase) {
 		return OP_OK;
 	}
 
-	// If we have no children, we can build the iterator now.
-	GraphContext *gc = QueryCtx_GetGraphCtx();
-	Schema *schema = GraphContext_GetSchema(gc, op->n.label, SCHEMA_NODE);
-	if(!schema) {
+	// resolve label ID at runtime
+	_update_label_id(op);
```
Comment (swilly22) [inline src/execution_plan/ops/op_node_by_label_scan.c:102]: might be redundant, we're already calling this function at construction time,
I think one of these calls is redundant. please check

File: src/execution_plan/ops/op_node_by_label_scan.c [hunk 8]
```diff
@@ -126,13 +132,10 @@ static inline void _ResetIterator(NodeByLabelScan *op) {
 		NodeID maxId = op->id_range->include_max ? op->id_range->max : op->id_range->max - 1 ;
 		RG_MatrixTupleIter_iterate_range(&op->iter, minId, maxId);
 	} else {
-		// id_range is NULL, this operation must have been freed previously
-		// rebuild the range iterator
-		GraphContext *gc = QueryCtx_GetGraphCtx();
-		Schema *schema = GraphContext_GetSchema(gc, op->n.label, SCHEMA_NODE);
-		if(!schema) return; // invalid schema, our consume function is NOP
+		_update_label_id(op);
```
Comment (swilly22) [inline src/execution_plan/ops/op_node_by_label_scan.c:135]: I feel like calling `_update_label_id` at runtime is redundant and should be removed, please validate this assumption.

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/execution_plan/ops/op_node_by_label_scan.c (hunk 7 above)
Comment (raz-mon) [reply src/execution_plan/ops/op_node_by_label_scan.c:102]: So we won't forget, writing here the example we started talking about:
`CREATE (:A) RETURN 1 UNION MATCH (a:A) RETURN a`

**Pull request merged**